File: apps/web/src/features/shared/standSync.mjs
// A seat write is never replayed. Read back the authoritative queue after
// either success or a lost acknowledgement before enabling another write.
// Polls started before a write cannot overwrite its reconciliation snapshot.
export class StandSync {
  constructor(readContext = () => null) {
    this.readContext = readContext;
    this.active = true;
    this.epoch = 0;
    this.sequence = 0;
    this.blocked = true;
    this.write = null;
  }
  activate() { this.active = true; this.epoch++; this.blocked = true; this.write = null; }
  dispose() { this.active = false; this.epoch++; this.write = null; }
  current(ticket) {
    return Boolean(ticket && this.active && ticket.epoch === this.epoch &&
      ticket.context === this.readContext());
  }
  beginRead({ reconcile = false } = {}) {
    if (!this.active || (this.write && !reconcile)) return null;
    return { epoch: this.epoch, context: this.readContext(), sequence: ++this.sequence };
  }
  currentRead(ticket) { return this.current(ticket) && ticket.sequence === this.sequence; }
  settleRead(ticket, success) {
    if (!this.currentRead(ticket)) return false;
    this.blocked = !success;
    return true;
  }
  beginWrite() {
    if (!this.active || this.blocked || this.write) return null;
    this.sequence++;
    this.blocked = true;
    this.write = { epoch: this.epoch, context: this.readContext() };
    return this.write;
  }
  currentWrite(ticket) { return this.current(ticket) && this.write === ticket; }
  finishWrite(ticket) {
    if (!this.currentWrite(ticket)) return false;
    this.write = null;
    return true;
  }
}

File: apps/mobile/smarttaxi_app/lib/features/shared/stand_sync.dart
/// Do not replay seat writes. Reconcile even a lost acknowledgement before
/// accepting another action; an older poll cannot overwrite that snapshot.
class StandSync {
  int _sequence = 0;
  int? _write;
  bool _disposed = false;
  bool blocked = true;

  int? beginRead({bool reconcile = false}) {
    if (_disposed || (_write != null && !reconcile)) return null;
    return ++_sequence;
  }

  bool currentRead(int ticket) => !_disposed && ticket == _sequence;

  bool settleRead(int ticket, bool success) {
    if (!currentRead(ticket)) return false;
    blocked = !success;
    return true;
  }

  int? beginWrite() {
    if (_disposed || blocked || _write != null) return null;
    blocked = true;
    return _write = ++_sequence;
  }

  bool currentWrite(int ticket) => !_disposed && _write == ticket;

  bool finishWrite(int ticket) {
    if (!currentWrite(ticket)) return false;
    _write = null;
    return true;
  }

  void dispose() {
    _disposed = true;
    _sequence++;
    _write = null;
  }
}
